src/backend/cuda/kernel/morph.hpp

```diff
@@ -239,22 +237,30 @@ void morph(Param<T> out, CParam<T> in, int windLen)
-        case 13: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,13>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
-        case 15: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,15>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
-        case 17: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,17>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
-        case 19: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,19>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
-        default: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 3>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  2: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  2>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  3: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  3>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  4: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  4>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  5: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  5>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  6: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  6>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  7: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  7>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  8: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  8>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case  9: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation,  9>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 10: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 10>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 11: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 11>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 12: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 12>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 13: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 13>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 14: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 14>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 15: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 15>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 16: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 16>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 17: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 17>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 18: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 18>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        case 19: CUDA_LAUNCH_SMEM((morphKernel<T, isDilation, 19>), blocks, threads, shrdSize, out, in, blk_x, blk_y); break;
+        default: CUDA_NOT_SUPPORTED(); break;
```

Comment:
We should document this limitation in doxygen. Also we should modify the CUDA_NOT_SUPPORTED function to say what is not supported.

Reply:
it would be better if we had a generic kernel which handled the cases we cannot address with template parameters.